Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,182 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+import SwiftParserDiagnostics
+
+actor DiagnosticReportManager {
+  /// A task that gets report from sourcekitd
+  private typealias ReportTask = Task<RelatedFullDocumentDiagnosticReport, Error>
+
+  private let sourcekitd: SourceKitD
+  private let syntaxTreeManager: SyntaxTreeManager
+  private let documentManager: DocumentManager
+  private let clientHasDiagnosticsCodeDescriptionSupport: Bool
+
+  private nonisolated var keys: sourcekitd_keys { return sourcekitd.keys }
+  private nonisolated var requests: sourcekitd_requests { return sourcekitd.requests }
+
+  /// The cache that stores reportTasks for snapshot id and buildSettings
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportTaskCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
+      buildSettings: SwiftCompileCommand?,
+      reportTask: ReportTask
+    )] = []
+
+  /// The number of reportTasks to keep
+  ///
+  /// - Note: This has been chosen without scientific measurements.
+  private let cacheSize = 5
+
+  init(
+    sourcekitd: SourceKitD,
+    syntaxTreeManager: SyntaxTreeManager,
+    documentManager: DocumentManager,
+    clientHasDiagnosticsCodeDescriptionSupport: Bool
+  ) {
+    self.sourcekitd = sourcekitd
+    self.syntaxTreeManager = syntaxTreeManager
+    self.documentManager = documentManager
+    self.clientHasDiagnosticsCodeDescriptionSupport = clientHasDiagnosticsCodeDescriptionSupport
+  }
+
+  func diagnosticReport(
+    for snapshot: DocumentSnapshot,
+    buildSettings: SwiftCompileCommand?
+  ) async throws -> RelatedFullDocumentDiagnosticReport {
+    if let report = try await reportTask(for: snapshot.id, buildSettings: buildSettings)?.value {
+      return report
+    }
```

Comment:
I would rewrite this slightly, just to make sure we only do an `await` if we already have a `reportTask`. That way you can verify that `diagnosticReport` runs up to its `return` statement without a suspension point (aka `await` call) and thus you don’t need to be concerned with high-level data races.

```suggestion
    if let reportTask = reportTask(for: snapshot.id, buildSettings: buildSettings) {
      return try await report.value
    }
```